src/mame/drivers/minitel_2_rpic.cpp

```diff
@@ -0,0 +1,475 @@
+			#endif
+			return port0;
+		case MCS51_PORT_P1:
+			#ifdef DEBUG_MINITEL
+			printf("port_r: read %02X from PORT1 - Keyboard -> %x\n", port1,((keyboard_x_row_reg>>7)&1));
+			#endif
+			return ( (port1&0xFE) | ((keyboard_x_row_reg>>7)&1) ) ;
+		case MCS51_PORT_P2:
+			#ifdef DEBUG_MINITEL
+			printf("port_r: read %02X from PORT2\n", port2);
+			#endif
+			return port2;
+		case MCS51_PORT_P3:
+			#ifdef DEBUG_MINITEL
+			printf("port_r: read %02X from PORT3\n", port3);
+			#endif
+			return port3;
+	}
+	return 0;
+}
+
+WRITE8_MEMBER(minitel_state::dev_crtl_reg_w)
+{
+	if( last_ctrl_reg != data)
+	{
+		#ifdef DEBUG_MINITEL
+		printf("minitel_state::hw_ctrl_reg : %x %x\n",offset, data);
+		#endif
+
+		if( (last_ctrl_reg ^ data) & CTRL_REG_DTMF )
+		{
+			#ifdef DEBUG_MINITEL
+			printf("CTRL_REG_DTMF : %d \n", data & CTRL_REG_DTMF );
+			#endif
+		}
+
+		if( (last_ctrl_reg ^ data) & CTRL_REG_MCBC )
+		{
+			#ifdef DEBUG_MINITEL
+			printf("CTRL_REG_MCBC : %d \n", data & CTRL_REG_MCBC );
+			#endif
+		}
+
+		if( (last_ctrl_reg ^ data) & CTRL_REG_OPTO )
+		{
+			#ifdef DEBUG_MINITEL
+			printf("CTRL_REG_OPTO : %d \n", data & CTRL_REG_OPTO );
+			#endif
+		}
+
+		if( (last_ctrl_reg ^ data) & CTRL_REG_RELAY )
+		{
+			#ifdef DEBUG_MINITEL
+			printf("CTRL_REG_RELAY : %d \n", data & CTRL_REG_RELAY );
+			#endif
+		}
+
+		if( (last_ctrl_reg ^ data) & CTRL_REG_CRTON )
+		{
+			#ifdef DEBUG_MINITEL
+			printf("CTRL_REG_CRTON : %d \n", data & CTRL_REG_CRTON );
+			#endif
+		}
+	}
+
+	last_ctrl_reg = data;
+}
+
+READ8_MEMBER(minitel_state::dev_keyb_ser_r)
+{
+	char kbdrow[8];
+
+	#ifdef DEBUG_MINITEL
+	printf("minitel_state::keyb read : %x\n",offset);
+	#endif
+
+	if ( keyboard_para_ser )
+	{
+		// load the 4014 with the keyboard row state
+		sprintf(kbdrow,"Y%.2X",offset>>8);
+		keyboard_x_row_reg = ioport(kbdrow)->read();
+		#ifdef DEBUG_MINITEL
+		printf("4014 Load : %s 0x%.2X\n",kbdrow,keyboard_x_row_reg);
+		#endif
+	}
+	else
+	{
+		//shift the keyboard register...
+		keyboard_x_row_reg = keyboard_x_row_reg << 1;
+	}
+
+	return 0xFF;
+}
+
+READ8_MEMBER ( minitel_state::ts9347_io_r )
+{
+	return m_ts9347->data_r(space, offset, 0xff);
+}
+
+WRITE8_MEMBER ( minitel_state::ts9347_io_w )
+{
+	#ifdef DEBUG_MINITEL
+	printf("minitel_state::ts9347_io_w : %x %x\n",offset, data);
+	#endif
+
+	m_ts9347->data_w(space, offset, data, 0xff);
+}
+
+TIMER_DEVICE_CALLBACK_MEMBER(minitel_state::minitel_scanline)
+{
+	m_ts9347->update_scanline((uint16_t)param);
+}
+
+static ADDRESS_MAP_START( mem_prg, AS_PROGRAM, 8, minitel_state )
+	AM_RANGE(0x0000, 0x7fff) AM_ROM
+ADDRESS_MAP_END
+
+static ADDRESS_MAP_START( mem_io, AS_IO, 8, minitel_state )
+	AM_RANGE(0x2000, 0x3fff) AM_READWRITE(dev_keyb_ser_r, dev_crtl_reg_w)
+	/* ts9347 */
+	AM_RANGE(0x4000, 0x5ffF) AM_READWRITE(ts9347_io_r, ts9347_io_w)
+	AM_RANGE(MCS51_PORT_P0, MCS51_PORT_P3) AM_READWRITE(port_r, port_w)
+ADDRESS_MAP_END
+
+/* Input ports */
+static INPUT_PORTS_START( minitel2 )
```

Comment:
What's the physical layout of the keyboard on this machine?  If it's AZERTY, the `A` key should be mapped to `KEYCODE_Q`, `Z` should be mapped to `KEYCODE_W`, etc.